tests/sqlAst/data/sql-ast-narrowing.php

```diff
@@ -440,6 +515,15 @@ public function joinSelectOutsideCondition(PDO $pdo): void
         assertType('PDOStatement<array{adaid: int<-32768, 32767>, 0: int<-32768, 32767>, eadavk: numeric-string|null, 1: numeric-string|null}>', $stmt);
     }
 
+    public function joinWhereCondition(PDO $pdo): void
+    {
+        $stmt = $pdo->query('SELECT c_json FROM typemix LEFT JOIN ada ON (c_int = adaid) WHERE c_json IS NOT NULL');
+        assertType('PDOStatement<array{c_json: string, 0: string}>', $stmt);
+
+        $stmt = $pdo->query('SELECT c_json FROM typemix RIGHT JOIN ada ON (c_int = adaid) WHERE c_json IS NOT NULL');
```

Comment:
Does this PR also support the case when the selected column which is narrowed via WHERE is from a joined table?
(Another test)

Reply:
Yes, it does. However, there's currently a bug in `QueryScope::narrowJoinedColumnType` where if the selected column is from the joined table but not in the join condition, then the AST type resolution returns `MixedType`, and so we revert to the query simulation type resolution (i.e. the WHERE condition is ignored, since query simulation doesn't narrow types like this).

Therefore, the following tests that I was planning to add would fail, because we get `string|null` (the query simulation type) instead of `string` (the expected AST type).
```php
$stmt = $pdo->query('SELECT c_json FROM ada LEFT JOIN typemix ON (c_int = adaid) WHERE c_json IS NOT NULL');
assertType('PDOStatement<array{c_json: string, 0: string}>', $stmt);

$stmt = $pdo->query('SELECT c_json FROM ada RIGHT JOIN typemix ON (c_int = adaid) WHERE c_json IS NOT NULL');
assertType('PDOStatement<array{c_json: string, 0: string}>', $stmt);
```
As a workaround, I've added `c_json` to the join condition in an unrealistic way (`c_json = c_json`) since no other table has a nullable string type, and I didn't want to make more schema changes (comparing against a non-nullable string column would prevent us from properly checking that the WHERE condition was used). This allows the AST type resolution to succeed, and we get the column from the joined table narrowed by the WHERE condition.

I'm working on a PR for the table join bug, but that won't be done until later this week I think.

Reply:
Cool thanks. Lets look at more cases in a separate PR. 

The PR as is - is already very valuable